### PR TITLE
Minimum PDF/A support

### PR DESCRIPTION
**Pull request by laurmaedje**
This is the absolute minimum of additions for PDF/A support (PDF/A-2b).

The API is not that great since there are various standard one can conform to and here it's just a bool, but maybe that is out of scope for the current svg2pdf architecture and should be revisited in the new architecture where Typst and svg2pdf share more code. 

Concretely, these changes should result in all rules [listed in the Typst tracking issue](https://github.com/typst/typst/issues/2942) being satisfied by svg2pdf. The only problems I could find were:

- Name length for font names: I adjusted the code to trim them.
- Maximum q nesting depth: I added a method to pdf-writer to track the depth (tracking it here would make the code uglier). I think that's generally useful. However, the validation itself happens in svg2pdf. When the nesting depth is too high, a new `TooMuchNesting` error is returned. In theory, we could use XObjects to reset the depth, but that is too much work right now.
- Notdef glyphs: When they are found and PDF/A support is enabled, a new `MissingGlyphs` error is returned.

Notably, only the notdef rule actually checks for PDF/A mode because the other rules turn out to apply to all PDF standards as far as I can tell!

I would release pdf-writer 0.11.1 before merging this.

### CODE REVIEW

File: src/render/text.rs [hunk 1]
```diff
@@ -518,18 +550,28 @@ pub fn fill_fonts(group: &Group, ctx: &mut Context, fontdb: &fontdb::Database) {
                             font.glyph_set.insert(g.id.0, g.text.clone());
                             font.glyph_remapper.remap(g.id.0);
                         }
+
+                        if ctx.options.pdfa && g.id.0 == 0 {
+                            return Err(ConversionError::MissingGlyphs);
+                        }
```
Comment (LaurenzV) [inline src/render/text.rs:556]: I don't remember, does PDF/A just talk about producing glyphs that are explicitly mapped to .notdef, or also implicitly (i.e. referencing a glyph that doesn't exist in the font)?

File: src/lib.rs [hunk 2]
```diff
@@ -148,6 +155,9 @@ pub struct ConversionOptions {
     ///
     /// _Default:_ `true`.
     pub embed_text: bool,
+
+    /// Whether to enforce PDF/A rules.
+    pub pdfa: bool,
```
Comment (LaurenzV) [inline src/lib.rs:160]: I would add a comment that this doesn't refer to making the whole file PDF/A compliant, though. From what I remember, there are other things missing, like adding certain metadata, which presumably will be implemented on the Typst side. 

But if you just use svg2pdf by itself, the resulting PDF won't be PDF/A compliant.

File: src/render/mask.rs [hunk 3]
```diff
@@ -34,7 +34,7 @@ pub fn create(
     let mut rc = ResourceContainer::new();
 
     let mut content = Content::new();
-    content.save_state();
+    content.save_state_checked()?;
```
Comment (LaurenzV) [inline src/render/mask.rs:37]: I presume you've done a Ctrl + F to make sure you caught all instances of `save_state`? 😄

**Review by LaurenzV — CHANGES_REQUESTED** [review]

**Comment by LaurenzV** [comment]
I think there are a few more points that require attention:
- Strings can't be longer than 32767 bytes -> Can't this happen if someone inputs a very long string? The text showing operator uses strings to display the glyphs, right?
- Can't contain more than 8388607 indirect objects -> I guess this can be done on the Typst side?
- 6.2.11.3.3: IIRC we don't write the WMode in the CMAP dictionary, should be an easy fix, though.
-

File: src/render/text.rs (hunk 1 above)
Comment (laurmaedje) [reply src/render/text.rs:556]: can that even happen?

File: src/lib.rs (hunk 2 above)
Comment (laurmaedje) [reply src/lib.rs:160]: True! I missed that.

File: src/render/mask.rs (hunk 3 above)
Comment (laurmaedje) [reply src/render/mask.rs:37]: I did

**Comment by laurmaedje** [comment]
- Strings: If I read the code correctly, you show each glyph with a separate invocation of the text showing operator
- Objects: Yep
- CMap: ah okay, good catch!

File: src/render/text.rs (hunk 1 above)
Comment (LaurenzV) [reply src/render/text.rs:556]: Fair point, I guess it can't assuming that usvg doesn't mess something up. :p

**Comment by LaurenzV** [comment]
> Strings: If I read the code correctly, you show each glyph with a separate invocation of the text showing operator

True, I forgot about that. 😅

File: src/render/text.rs [hunk 4]
```diff
@@ -367,17 +376,36 @@ pub fn render(
     Ok(())
 }
 
+/// Creates the base font name for a font with a specific glyph subset.
+/// Consists of a subset tag and the PostScript name of the font.
+///
+/// Returns a string of length maximum 116, so that even with `-Identity-H`
+/// added it does not exceed the maximum PDF/A name length of 127.
```
Comment (LaurenzV) [inline src/render/text.rs:383]: Can font names contain binary characters, in theory? If so, they will be escaped in `pdf-writer`.

Comment (laurmaedje) [reply src/render/text.rs:383]: My reading of the spec is that the 127 bytes limit is about the resulting name and not the encoding as PDF bytes. Would need to check with verapdf to be sure.

Comment (laurmaedje) [reply src/render/text.rs:383]: I checked and it is indeed fine. `Name("(".repeat(100).as_bytes())` is accepted by verapdf.

**Review by LaurenzV — APPROVED** [review]

**Pull request merged**